diskless master, avoid bgsave child hung when fork parent crashes

oranagra opened the pull request:
During a diskless sync, if the master main process crashes, the child would
have hung in `write`. This fix closes the read fd on the child side, so that if the
parent crashes, the child will get a write error and exit.

This change also fixes disk-based replication, BGSAVE and AOFRW.
In that case the child wouldn't have been hang, it would have just kept
running until done which may be pointless.

There is a certain degree of risk here. in case there's a BGSAVE child that could
maybe succeed and the parent dies for some reason, the old code would have let
the child keep running and maybe succeed and avoid data loss.
On the other hand, if the parent is restarted, it would have loaded an old rdb file
(or none), and then the child could reach the end and rename the rdb file (data
conflicting with what the parent has), or also have a race with another BGSAVE
child that the new parent started.

Note that i removed a comment saying a write error will be ignored in the child
and handled by the parent (this comment was very old and i don't think relevant).

yossigo reviewed (approved):
@oranagra LGTM, do you think we need a test for that?

oranagra commented:
i don't think writing a test for this case worth the effort, i imagine it'll be complicated, but let me try and see.

oranagra commented:
@yossigo i added a test, and decided to fix a similar problem with disk-based replication.
in that case the child wouldn't have been hang, it would have just kept running until done which may be pointless.
this fix also affects BGSAVE and AOFRW.

note that i removed a comment saying a write error will be ignored in the child and handled by the parent (this comment was very old).
there is a certain degree of risk here. in case there's a bgsave child that could maybe succeed and the parent dies for some reason, the old code would have let the child keep running and maybe avoid data loss.
on the other hand, if the parent is restarted, it would have loaded an old rdb file (or none), and then the child could reach the end and rename the rdb file (data conflicting with what the parent has), or also have a race with another BGSAVE child that's started.

WDYT?

yossigo reviewed (approved):
@oranagra I think this change makes sense. It's better to give up the opportunistic benefit of a crash data recovery during disk-based replication to avoid race conditions with a runaway child, IMHO.

enjoy-binbin commented on tests/integration/replication.tcl:
```diff
@@ -1012,6 +1012,45 @@ test "diskless replication child being killed is collected" {
+                wait_for_condition 500 10 {
+                    [process_is_alive $fork_child_pid] == 0
+                } else {
+                    fail "rdb child didn't terminate"
```
look like the test break all the daily.
https://github.com/redis/redis/actions/runs/3432845692/jobs/5722529615

change the `wait_for_condition 500 10 {` to `wait_for_condition 5000 10 {` for test:
https://github.com/enjoy-binbin/redis/actions/runs/3433763843

But I do see that the test takes a very short time in some other daily tests, just a try
```
[ok]: replication dies when parent is killed - diskless: yes (1109 ms)
[ok]: replication dies when parent is killed - diskless: no (1689 ms)
```

oops, the increase didn't work

oranagra replied:
i think the problem is with `ps -p` not working on some systems

oranagra replied:
scrap that, i see it's used in instances.tcl

oranagra replied:
or maybe it is `ps -p`, i see it fails on all centos and alpine tests, and i see we don't run the cluster and sentinel tests on these...

oranagra replied:
scrap that. another wrong observation (we do run the cluster tests on these).
still it seems some issue related to centos and alpine

enjoy-binbin replied:
maybe the logs can help you
```
1370:S 10 Nov 2022 14:03:57.552 * Before turning into a replica, using my own master parameters to synthesize a cached master: I may be able to synchronize with the new master with just a partial transfer.
1370:S 10 Nov 2022 14:03:57.552 * Connecting to MASTER 127.0.0.1:21113
1370:S 10 Nov 2022 14:03:57.552 * MASTER <-> REPLICA sync started
1370:S 10 Nov 2022 14:03:57.552 * REPLICAOF 127.0.0.1:21113 enabled (user request from 'id=4 addr=127.0.0.1:46529 laddr=127.0.0.1:21114 fd=8 name= age=0 idle=0 flags=N db=9 sub=0 psub=0 ssub=0 multi=-1 qbuf=45 qbuf-free=16339 argv-mem=23 multi-mem=0 rbs=16384 rbp=16384 obl=0 oll=0 omem=0 tot-mem=33581 events=r cmd=replicaof user=default redir=-1 resp=2')
1370:S 10 Nov 2022 14:03:57.552 * Non blocking connect for SYNC fired the event.
1370:S 10 Nov 2022 14:03:57.553 * Master replied to PING, replication can continue...
1370:S 10 Nov 2022 14:03:57.553 * Trying a partial resynchronization (request 8620cff8d9a8397c7c512fb182e8a5ab3e121c36:1).
1370:S 10 Nov 2022 14:03:57.554 * Full resync from master: f3e2063538e718b8e2f112b3e3e255529fad924c:0
1370:S 10 Nov 2022 14:03:58.669 # I/O error reading bulk count from MASTER: Invalid argument
1370:S 10 Nov 2022 14:03:58.669 * Reconnecting to MASTER 127.0.0.1:21113 after failure
1370:S 10 Nov 2022 14:03:58.669 * MASTER <-> REPLICA sync started
1370:S 10 Nov 2022 14:03:58.669 # Error condition on socket for SYNC: Connection refused
1370:S 10 Nov 2022 14:03:59.444 * Connecting to MASTER 127.0.0.1:21113
1370:S 10 Nov 2022 14:03:59.444 * MASTER <-> REPLICA sync started
1370:S 10 Nov 2022 14:03:59.444 # Error condition on socket for SYNC: Connection refused
1370:S 10 Nov 2022 14:04:00.446 * Connecting to MASTER 127.0.0.1:21113
1370:S 10 Nov 2022 14:04:00.446 * MASTER <-> REPLICA sync started
1370:S 10 Nov 2022 14:04:00.446 # Error condition on socket for SYNC: Connection refused
1370:S 10 Nov 2022 14:04:01.449 * Connecting to MASTER 127.0.0.1:21113
1370:S 10 Nov 2022 14:04:01.449 * MASTER <-> REPLICA sync started
1370:S 10 Nov 2022 14:04:01.449 # Error condition on socket for SYNC: Connection refused
1370:S 10 Nov 2022 14:04:02.452 * Connecting to MASTER 127.0.0.1:21113
1370:S 10 Nov 2022 14:04:02.452 * MASTER <-> REPLICA sync started
1370:S 10 Nov 2022 14:04:02.452 # Error condition on socket for SYNC: Connection refused
1370:S 10 Nov 2022 14:04:03.454 * Connecting to MASTER 127.0.0.1:21113
1370:S 10 Nov 2022 14:04:03.454 * MASTER <-> REPLICA sync started
1370:S 10 Nov 2022 14:04:03.454 # Error condition on socket for SYNC: Connection refused
1370:S 10 Nov 2022 14:04:03.922 - Client closed connection id=4 addr=127.0.0.1:46529 laddr=127.0.0.1:21114 fd=8 name= age=6 idle=6 flags=N db=9 sub=0 psub=0 ssub=0 multi=-1 qbuf=0 qbuf-free=16384 argv-mem=0 multi-mem=0 rbs=1024 rbp=0 obl=0 oll=0 omem=0 tot-mem=18174 events=r cmd=replicaof user=default redir=-1 resp=2
1370:signal-handler (1668089043) Received SIGTERM scheduling shutdown...
1370:S 10 Nov 2022 14:04:03.956 # User requested shutdown...
1370:S 10 Nov 2022 14:04:03.956 * Saving the final RDB snapshot before exiting.
1370:S 10 Nov 2022 14:04:03.958 * DB saved on disk
1370:S 10 Nov 2022 14:04:03.958 * Removing the pid file.
1370:S 10 Nov 2022 14:04:03.958 * Removing the unix socket file.
1370:S 10 Nov 2022 14:04:03.958 # Redis is now ready to exit, bye bye...








1357:M 10 Nov 2022 14:03:57.412 - Accepted 127.0.0.1:44229
1357:M 10 Nov 2022 14:03:57.412 - Client closed connection id=3 addr=127.0.0.1:44229 laddr=127.0.0.1:21113 fd=8 name= age=0 idle=0 flags=N db=0 sub=0 psub=0 ssub=0 multi=-1 qbuf=0 qbuf-free=16384 argv-mem=0 multi-mem=0 rbs=16384 rbp=16384 obl=0 oll=0 omem=0 tot-mem=33534 events=r cmd=ping user=default redir=-1 resp=2
1357:M 10 Nov 2022 14:03:57.415 - Accepted 127.0.0.1:41103
1357:M 10 Nov 2022 14:03:57.552 - Accepted 127.0.0.1:33158
1357:M 10 Nov 2022 14:03:57.553 * Replica 127.0.0.1:21114 asks for synchronization
1357:M 10 Nov 2022 14:03:57.553 * Partial resynchronization not accepted: Replication ID mismatch (Replica asked for '8620cff8d9a8397c7c512fb182e8a5ab3e121c36', my replication IDs are '1df0435ecc811017d46dc24af26ffbf068604ecc' and '0000000000000000000000000000000000000000')
1357:M 10 Nov 2022 14:03:57.553 * Replication backlog created, my new replication IDs are 'f3e2063538e718b8e2f112b3e3e255529fad924c' and '0000000000000000000000000000000000000000'
1357:M 10 Nov 2022 14:03:57.553 * Starting BGSAVE for SYNC with target: disk
1357:M 10 Nov 2022 14:03:57.553 * Background saving started by pid 1380
1380:C 10 Nov 2022 14:03:57.555 - Fork CoW for RDB: current 0 MB, peak 0 MB, average 0 MB
1380:C 10 Nov 2022 14:03:58.668 - Fork CoW for RDB: current 2 MB, peak 2 MB, average 1 MB
1380:C 10 Nov 2022 14:03:58.668 # Child failed reporting info to parent, exiting. Broken pipe
```

enjoy-binbin replied:
odd, but it works on my machine (I even ran a loop), centos 7.5
```
[root@binblog redis]# cat /etc/redhat-release
CentOS Linux release 7.5.1804 (Core)

[ok]: replication dies when parent is killed - diskless: yes (1059 ms)
[ok]: replication dies when parent is killed - diskless: no (1252 ms)
```

enjoy-binbin replied:
ok, i see those child are both in defunct state (ps -p result)
```
4281 is the parent, 4306 is the child


    PID TTY          TIME CMD
   4281 ?        00:00:00 redis-server
    PID TTY          TIME CMD

    kill the parent

   4281 ?        00:00:00 redis-server <defunct>
    PID TTY          TIME CMD

    wait_for_condition parent die

    wait_for_condition checkint the child
   4306 ?        00:00:00 redis-server
    PID TTY          TIME CMD
   4306 ?        00:00:00 redis-server
    PID TTY          TIME CMD
   4306 ?        00:00:00 redis-server <defunct>
    PID TTY          TIME CMD
   4306 ?        00:00:00 redis-server <defunct>
```

for the record, i did see the child call "exit(1)" because it print the `Child failed reporting info to parent, exiting. Broken pipe` log

oranagra replied:
interesting..  that's actually right that they should stay defunct since the parent was killed before collecting them with `wait`, but that's actually all we can hope for.

so i guess the next question is: why doesn't it happen on Ubuntu as well, and why doesn't it happen on your CentOS?
the next question would be how we disregard them in the test (grep on `ps` output)?

@yossigo please chime in.

yossigo replied:
@oranagra I think the issue is not CentOS/Alpine but the fact that we're running those tests in a container, and don't have an `init(8)` process (i.e. systemd) to reap zombies.

We need `process_is_alive` to filter these processes by their `<defunct>` or `Z` state.

enjoy-binbin replied:
a simple way to pass our daily (I tested it, it works in all tests, and i need to go to bed, take it from here):
```diff
proc process_is_alive pid {
    if {[catch {exec ps -p $pid} err]} {
        return 0
    } else {
+        if {[string match "*<defunct>*" $err]} { return 0 }
        return 1
    }
}
```

oranagra replied:
thanks. make a PR with that and some other adjustments.
